### PR TITLE
Tasks and Subject relationship

### PR DESCRIPTION
**Pull request by leduytungframgia**

### CODE REVIEW

File: app/controllers/supervisor/subjects_controller.rb [hunk 1]
```diff
@@ -19,15 +21,28 @@ def show
     @subject = Subject.find params[:id]
   end
 
+  def update
+    @subject = Subject.find(params[:id])
```
Comment (NguyenTanDuc) [inline app/controllers/supervisor/subjects_controller.rb]: remove  ()

File: app/views/subjects/show.html.erb [hunk 2]
```diff
@@ -1,15 +0,0 @@
-<% provide :title, @subject.name %>
```
Comment (dieunb) [inline app/views/subjects/show.html.erb]: checkout this file

File: app/views/supervisor/subjects/_form.html.erb [hunk 3]
```diff
@@ -5,8 +5,46 @@
...
   <%= f.text_area :instruction, class: 'form-control' %>
   <%= f.label :start_date, t(:start_date) %>
-  <%= f.password_field :start_date, class: 'form-control' %>
+  <%= f.text_field :start_date, class: 'form-control' %>
   <%= f.label :end_date, t(:end_date) %>
-  <%= f.password_field :end_date, class: 'form-control' %>
-  <%= f.submit "Submit", class: "btn btn-primary" %>
+  <%= f.text_field :end_date, class: 'form-control' %>
+  
+  <% n = 0 %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: remove

File: app/views/supervisor/subjects/_form.html.erb [hunk 4]
```diff
@@ -5,8 +5,46 @@
...
+  <%= f.fields_for :tasks do |ff| %>
+        <% n = n + 1 %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: remove

File: app/views/supervisor/subjects/_form.html.erb [hunk 5]
```diff
@@ -5,8 +5,46 @@
...
+        <h2>
+          Task <%= n %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: replace `n` by `ff.index + 1`

File: app/views/supervisor/subjects/_form.html.erb [hunk 6]
```diff
@@ -5,8 +5,46 @@
...
+  <%= f.text_field :start_date, class: 'form-control' %>
   <%= f.label :end_date, t(:end_date) %>
-  <%= f.password_field :end_date, class: 'form-control' %>
-  <%= f.submit "Submit", class: "btn btn-primary" %>
+  <%= f.text_field :end_date, class: 'form-control' %>
+  
+  <% n = 0 %>
+  <%= f.fields_for :tasks do |ff| %>
+        <% n = n + 1 %>
+        <h2>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: app/views/supervisor/subjects/_form.html.erb [hunk 7]
```diff
@@ -5,8 +5,46 @@
...
+          Task <%= n %>
+        </h2>
+          <%= ff.label :title %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: app/views/supervisor/subjects/_form.html.erb [hunk 8]
```diff
@@ -5,8 +5,46 @@
...
+          <%= ff.text_field :title, class:'input_fn' %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: app/views/supervisor/subjects/_form.html.erb [hunk 9]
```diff
@@ -5,8 +5,46 @@
...
+          <%= ff.label :content %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: app/views/supervisor/subjects/_form.html.erb [hunk 10]
```diff
@@ -5,8 +5,46 @@
...
+          <%= ff.text_area :content, class:'input_ta' %>
+          <%= ff.label :start_date %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: app/views/supervisor/subjects/_form.html.erb [hunk 11]
```diff
@@ -5,8 +5,46 @@
...
+  <% n = 0 %>
+  <%= f.fields_for :tasks do |ff| %>
+        <% n = n + 1 %>
+        <h2>
+          Task <%= n %>
+        </h2>
+          <%= ff.label :title %>
+          <%= ff.text_field :title, class:'input_fn' %>
+          <%= ff.label :content %>
+          <%= ff.text_area :content, class:'input_ta' %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: app/views/supervisor/subjects/_form.html.erb [hunk 12]
```diff
@@ -5,8 +5,46 @@
...
+          <%= ff.label :start_date %>
+          <%= ff.text_field :start_date %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: app/views/supervisor/subjects/_form.html.erb [hunk 13]
```diff
@@ -5,8 +5,46 @@
...
+          <%= ff.label :end_date %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: app/views/supervisor/subjects/_form.html.erb [hunk 14]
```diff
@@ -5,8 +5,46 @@
...
+          <%= ff.text_field :end_date%>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: app/views/supervisor/subjects/_form.html.erb [hunk 15]
```diff
@@ -5,8 +5,46 @@
...
+  <% end %>
+
+  <div class="row">
+  <div class="col-md-6 col-md-offset-3">
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: indent

File: config/routes.rb [hunk 16]
```diff
@@ -9,8 +9,10 @@
 
   namespace :supervisor do
     resources :users
-    resources :subjects
+    resources :subjects do
+        resources :tasks
```
Comment (dieunb) [inline config/routes.rb]: indent

File: app/models/subject.rb [hunk 17]
```diff
@@ -2,4 +2,9 @@ class Subject < ActiveRecord::Base
   has_many :subject_courses
   has_many :courses, through: :subject_courses
   has_many :tasks
```
Comment (dieunb) [inline app/models/subject.rb]: `dependent: :destroy`

File: app/models/subject.rb [hunk 18]
```diff
@@ -2,4 +2,9 @@ class Subject < ActiveRecord::Base
...
+  
+  validates :name, presence: true
+  validates :instruction, presence: true
+
+  accepts_nested_attributes_for :tasks
```
Comment (dieunb) [inline app/models/subject.rb]: `reject if`

File: app/models/subject.rb [hunk 19]
```diff
@@ -1,5 +1,16 @@
...
   has_many :subject_courses
   has_many :courses, through: :subject_courses
-  has_many :tasks
+  has_many :tasks, dependent: :destroy
+  
+  validates :name, presence: true
+  validates :instruction, presence: true
+
+  accepts_nested_attributes_for :tasks, reject_if: :reject_posts, allow_destroy: true
+  def reject_posts attributed
```
Comment (dieunb) [inline app/models/subject.rb]: private methods

File: app/models/task.rb [hunk 20]
```diff
@@ -1,2 +1,5 @@
 class Task < ActiveRecord::Base
+  belongs_to :subject
+  validates :title, presence: true
```
Comment (dieunb) [inline app/models/task.rb]: add new white line above this

File: app/views/supervisor/subjects/_form.html.erb [hunk 21]
```diff
@@ -5,8 +5,42 @@
...
   <%= f.label :start_date, t(:start_date) %>
-  <%= f.password_field :start_date, class: 'form-control' %>
+  <%= f.text_field :start_date, class: 'form-control' %>
   <%= f.label :end_date, t(:end_date) %>
-  <%= f.password_field :end_date, class: 'form-control' %>
-  <%= f.submit "Submit", class: "btn btn-primary" %>
+  <%= f.text_field :end_date, class: 'form-control' %>
+
+  <%= f.fields_for :tasks do |ff| %>
+    <h2>Task <%= ff.index +1 %></h2>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: missing a space after `+`

File: app/views/supervisor/subjects/_form.html.erb [hunk 22]
```diff
@@ -5,8 +5,42 @@
...
+    <%= ff.label :start_date %>
+    <%= ff.text_field :start_date %>
+    <%= ff.label :end_date %>
+    <%= ff.text_field :end_date%>
+  <% end %>
+
+  <div class="row">
+    <div class="col-md-6 col-md-offset-3">
+      <%= f.fields_for :tasks, @subject.tasks.build do |fff| %>
+      <%= render 'shared/error_messages', object: fff.object %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: remove, no need

File: app/models/subject.rb [hunk 23]
```diff
@@ -1,5 +1,18 @@
...
+  has_many :tasks, dependent: :destroy
+  
+  validates :name, presence: true
+  validates :instruction, presence: true
+
+  accepts_nested_attributes_for :tasks, reject_if: :reject_posts, allow_destroy: true
+  
+  private
+  def reject_posts attributed
+    attributed['title'].blank?
```
Comment (dieunb) [inline app/models/subject.rb]: ý của em là như thế này phải không?
```
attributed['title'].blank? || attributed['content'].blank? || attributed['start_date'].blank? || attributed['end_date'].blank?
```

File: app/helpers/application_helper.rb [hunk 24]
```diff
@@ -8,4 +7,16 @@ def full_title page_title = ''
...
+    f.hidden_field(:_destroy) + link_to_function(name, "remove_fields(this)")
+  end
+  
+  def link_to_add_fields(name, f, association)
+    new_object = f.object.class.reflect_on_association(association).klass.new
+    fields = f.fields_for(association, new_object, :child_index => "new_#{association}") do |builder|
+      render(association.to_s.singularize + "_fields", :f => builder)
+    end
+    link_to name, "#", 
+    :onclick => h("add_fields(this, \"#{association}\", \"#{escape_javascript(fields)}\")"), remote: true
```
Comment (NguyenTanDuc) [inline app/helpers/application_helper.rb]: replace =>

File: app/views/supervisor/subjects/show.html.erb [hunk 25]
```diff
@@ -1,14 +1,16 @@
...
+  <div class="alert alert-info" role="alert"><span class="user">
     <h3><%= link_to @subject.name, @subject %></h3>
     <span class="timestamp">
-      Posted <%= time_ago_in_words @subject.created_at %> ago.
-      <%= link_to "Delete", @subject, method: :delete,
-                                data: {confirm: t(:confirm_notification)} %>
+    Posted <%= time_ago_in_words @subject.created_at %> ago.
+    <%= link_to "Delete", supervisor_subject_path, method: :delete,
+                                  data: {confirm: t(:confirm_notification)} %>
+    <%= link_to "Edit subject", edit_supervisor_subject_path(@subject) %>
```
Comment (NguyenTanDuc) [inline app/views/supervisor/subjects/show.html.erb]: I18n

File: app/views/supervisor/subjects/show.html.erb [hunk 26]
```diff
@@ -1,14 +1,16 @@
...
 <div class="panel panel-info">
-  <div class="alert alert-info" role="alert">
+  <div class="alert alert-info" role="alert"><span class="user">
     <h3><%= link_to @subject.name, @subject %></h3>
     <span class="timestamp">
-      Posted <%= time_ago_in_words @subject.created_at %> ago.
-      <%= link_to "Delete", @subject, method: :delete,
-                                data: {confirm: t(:confirm_notification)} %>
+    Posted <%= time_ago_in_words @subject.created_at %> ago.
+    <%= link_to "Delete", supervisor_subject_path, method: :delete,
```
Comment (NguyenTanDuc) [inline app/views/supervisor/subjects/show.html.erb]: I18n

File: app/helpers/application_helper.rb [hunk 27]
```diff
@@ -1,5 +1,4 @@
 module ApplicationHelper
-
   def full_title page_title = ''
     base_title = "Welcome to Education App"
```
Comment (dieunb) [inline app/helpers/application_helper.rb]: i18n

File: app/helpers/application_helper.rb [hunk 28]
```diff
@@ -8,4 +7,16 @@ def full_title page_title = ''
       "#{page_title} | #{base_title}"
     end
   end
+  def link_to_remove_fields(name, f)
```
Comment (dieunb) [inline app/helpers/application_helper.rb]: new line and remove `()`

File: app/helpers/application_helper.rb [hunk 29]
```diff
@@ -8,4 +7,16 @@ def full_title page_title = ''
...
+    f.hidden_field(:_destroy) + link_to_function(name, "remove_fields(this)")
+  end
+  
+  def link_to_add_fields(name, f, association)
```
Comment (dieunb) [inline app/helpers/application_helper.rb]: remove `()`

File: app/helpers/application_helper.rb [hunk 30]
```diff
@@ -8,4 +7,16 @@ def full_title page_title = ''
...
+    new_object = f.object.class.reflect_on_association(association).klass.new
+    fields = f.fields_for(association, new_object, :child_index => "new_#{association}") do |builder|
```
Comment (dieunb) [inline app/helpers/application_helper.rb]: coding convention

File: app/helpers/application_helper.rb [hunk 31]
```diff
@@ -8,4 +7,16 @@ def full_title page_title = ''
...
+      render(association.to_s.singularize + "_fields", :f => builder)
```
Comment (dieunb) [inline app/helpers/application_helper.rb]: coding convention

File: app/helpers/subjects_helper.rb [hunk 32]
```diff
@@ -1,2 +0,0 @@
-module SubjectsHelper
```
Comment (dieunb) [inline app/helpers/subjects_helper.rb]: checkout this file

File: app/views/supervisor/subjects/_form.html.erb [hunk 33]
```diff
@@ -5,8 +5,12 @@
...
   <%= f.text_area :instruction, class: 'form-control' %>
   <%= f.label :start_date, t(:start_date) %>
-  <%= f.password_field :start_date, class: 'form-control' %>
+  <%= f.text_field :start_date, class: 'form-control' %>
   <%= f.label :end_date, t(:end_date) %>
-  <%= f.password_field :end_date, class: 'form-control' %>
-  <%= f.submit "Submit", class: "btn btn-primary" %>
+  <%= f.text_field :end_date, class: 'form-control' %>
+  <%= f.fields_for :tasks do |builder| %>
+    <%= render "task_fields", :f => builder %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_form.html.erb]: coding convention

File: app/views/supervisor/subjects/_task_fields.html.erb [hunk 34]
```diff
@@ -0,0 +1,9 @@
+<h2>Task </h2>
+  <%= f.label :title %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_task_fields.html.erb]: indent

File: app/assets/javascripts/application_jquery.js [hunk 35]
```diff
@@ -0,0 +1,10 @@
+function remove_fields(link) {
```
Comment (dieunb) [inline app/assets/javascripts/application_jquery.js]: duplicated code

File: app/assets/javascripts/application_jquery.js [hunk 36]
```diff
@@ -0,0 +1,10 @@
...
+  $(link).prev("input[type=hidden]").val("1");
+  $(link).closest(".fields").hide();
+}
+
+function add_fields(link, association, content) {
```
Comment (dieunb) [inline app/assets/javascripts/application_jquery.js]: duplicated code

File: app/helpers/application_helper.rb [hunk 37]
```diff
@@ -1,11 +1,22 @@
...
   def full_title page_title = ''
-    base_title = "Welcome to Education App"
+    base_title = t(:welcome)
     if page_title.empty?
       base_title
     else
       "#{page_title} | #{base_title}"
     end
   end
+  def link_to_remove_fields name, f
```
Comment (dieunb) [inline app/helpers/application_helper.rb]: add a new white line above this

File: app/views/supervisor/subjects/_task_fields.html.erb [hunk 38]
```diff
@@ -0,0 +1,9 @@
+<h2> Task </h2>
+  <%= f.label :title %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_task_fields.html.erb]: indent

File: app/views/supervisor/subjects/_task_fields.html.erb [hunk 39]
```diff
@@ -0,0 +1,9 @@
...
+  <%= f.text_field :title %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_task_fields.html.erb]: indent

File: app/views/supervisor/subjects/_task_fields.html.erb [hunk 40]
```diff
@@ -0,0 +1,9 @@
...
+  <%= f.label :content %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_task_fields.html.erb]: indent

File: app/views/supervisor/subjects/_task_fields.html.erb [hunk 41]
```diff
@@ -0,0 +1,9 @@
...
+  <%= f.text_area :content %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_task_fields.html.erb]: indent

File: app/views/supervisor/subjects/_task_fields.html.erb [hunk 42]
```diff
@@ -0,0 +1,9 @@
...
+  <%= f.label :start_date %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_task_fields.html.erb]: indent

File: app/views/supervisor/subjects/_task_fields.html.erb [hunk 43]
```diff
@@ -0,0 +1,9 @@
...
+  <%= f.text_field :start_date %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_task_fields.html.erb]: indent

File: app/views/supervisor/subjects/_task_fields.html.erb [hunk 44]
```diff
@@ -0,0 +1,9 @@
...
+  <%= f.label :end_date %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_task_fields.html.erb]: indent

File: app/views/supervisor/subjects/_task_fields.html.erb [hunk 45]
```diff
@@ -0,0 +1,9 @@
...
+  <%= f.text_field :end_date %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_task_fields.html.erb]: indent

File: app/views/supervisor/subjects/show.html.erb [hunk 46]
```diff
@@ -1,14 +1,16 @@
 <% provide :title, @subject.name %>
 <div class="panel panel-info">
-  <div class="alert alert-info" role="alert">
+  <div class="alert alert-info" role="alert"><span class="user">
     <h3><%= link_to @subject.name, @subject %></h3>
     <span class="timestamp">
-      Posted <%= time_ago_in_words @subject.created_at %> ago.
-      <%= link_to "Delete", @subject, method: :delete,
-                                data: {confirm: t(:confirm_notification)} %>
+    Posted <%= time_ago_in_words @subject.created_at %> ago.
```
Comment (dieunb) [inline app/views/supervisor/subjects/show.html.erb]: indent

File: app/views/supervisor/subjects/show.html.erb [hunk 47]
```diff
@@ -1,14 +1,16 @@
...
+    <%= link_to t(:delete_link), supervisor_subject_path, method: :delete,
```
Comment (dieunb) [inline app/views/supervisor/subjects/show.html.erb]: indent

File: app/views/supervisor/subjects/show.html.erb [hunk 48]
```diff
@@ -1,14 +1,16 @@
...
+                                  data: {confirm: t(:confirm_notification)} %>
+    <%= link_to t(:edit_subject), edit_supervisor_subject_path(@subject) %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/show.html.erb]: indent

File: app/assets/javascripts/application_jquery.js (hunk 35 above)
Comment (leduytungframgia) [inline app/assets/javascripts/application_jquery.js]: fixed

File: app/assets/javascripts/application_jquery.js (hunk 36 above)
Comment (leduytungframgia) [inline app/assets/javascripts/application_jquery.js]: fixed

**Pull request closed without merging**